src/trilateration/trilateration.cpp

```diff
@@ -0,0 +1,186 @@
+
+    const double b_x = delta_x / pow(delta, 2) * (pow(delta, 2) -
+            pow(delta_x, 2));
+    const double b_y = delta_y / pow(epsilon, 2) * (pow(epsilon, 2) -
+            pow(delta_y, 2));
+    const double b_z = delta_z / pow(zeta, 2) * (pow(zeta, 2) -
+            pow(delta_z, 2));
+
+    const double c_x = pow((pow(delta_x, 2) - pow(delta, 2)) / (2 * delta), 2);
+    const double c_y = pow((pow(delta_y, 2) - pow(epsilon, 2)) / (2 * epsilon),
+            2);
+    const double c_z = pow((pow(delta_z, 2) - pow(zeta, 2)) / (2 * zeta), 2);
+
+    const double a = a_x + a_y + a_z - 1;
+    const double b = b_x + b_y + b_z;
+    const double c = c_x + c_y + c_z;
+
+    /*
+     * Find the determinant of the quadratic to ensure that it is not an
+     * imaginary solution.
+     */
+    const double determinant = pow(b, 2) - 4 * a * c;
+    if (determinant < 0)
+    {
+        ROS_INFO_STREAM("dX = " << delta_x);
+        ROS_INFO_STREAM("dY = " << delta_y);
+        ROS_INFO_STREAM("dZ = " << delta_z);
+        ROS_INFO_STREAM("a: " << a);
+        ROS_INFO_STREAM("b: " << b);
+        ROS_INFO_STREAM("c: " << c);
+        ROS_INFO_STREAM("Determinant: " << determinant);
+        ROS_ERROR_STREAM("The trilateration encountered an imaginary "
+                "solution.");
+        return;
+    }
+
+    /*
+     * Calculate the distance from the pinger for the two possible solutions of
+     * the quadratic equation.
+     */
+    const double p_zero_solution_one = (-1*b + sqrt(determinant)) / (2 * a);
+    const double p_zero_solution_two = (-1*b - sqrt(determinant)) / (2 * a);
+
+    /*
+     * Calculate the two potential position readings.
+     */
+    const double pos_x_one = (delta_x * (2 * p_zero_solution_one - delta_x) +
+            delta * delta) / (2 * delta);
+    const double pos_x_two = (delta_x * (2 * p_zero_solution_two - delta_x) +
+            delta * delta) / (2 * delta);
+    const double pos_y_one = (delta_y * (2 * p_zero_solution_one - delta_y) +
+            epsilon * epsilon) / (2 * epsilon);
+    const double pos_y_two = (delta_y * (2 * p_zero_solution_two - delta_y) +
+            epsilon * epsilon) / (2 * epsilon);
+    const double pos_z_one = (delta_z * (2 * p_zero_solution_one - delta_z) +
+            zeta * zeta) / (2 * zeta);
+    const double pos_z_two = (delta_z * (2 * p_zero_solution_two - delta_z) +
+            zeta * zeta) / (2 * zeta);
+
+    /*
+     * Construct and publish the position message.
+     */
+    robosub::Position position_msg;
+    position_msg.header.stamp = ros::Time::now();
+    position_msg.distance_position_one = p_zero_solution_one;
+    position_msg.position_one.x = pos_x_one;
+    position_msg.position_one.y = pos_y_one;
+    position_msg.position_one.z = pos_z_one;
+    position_msg.distance_position_two = p_zero_solution_two;
+    position_msg.position_two.x = pos_x_two;
+    position_msg.position_two.y = pos_y_two;
+    position_msg.position_two.z = pos_z_two;
```

Comment:
Convert the position to be absolute in the global perspective and to utilize the pinger as the origin position. Currently, the  position is the location of the pinger relative to the sub's orientation. Subscribe to the orientation topic and rotate the coordinates of the pinger to be unrelated to the submarine's orientation.